File: QubitServer/src/main/java/org/labrad/qubits/channeldata/AnalogDataTime.java
package org.labrad.qubits.channeldata;

import java.util.concurrent.Future;

import org.labrad.qubits.channels.AnalogChannel;
import org.labrad.qubits.proxies.DeconvolutionProxy;
import org.labrad.qubits.util.Futures;

import com.google.common.base.Function;
import com.google.common.base.Preconditions;

public class AnalogDataTime extends AnalogDataBase {

  private double[] rawData = null;
  private boolean averageEnds;
  private boolean dither;
  private int[] deconvolvedData = null;

  public AnalogDataTime(double[] data, boolean isDeconvolved, boolean averageEnds, boolean dither) {
    this.rawData = data;
    this.averageEnds = averageEnds;
    this.dither = dither;
    if (isDeconvolved) {
      int[] values = new int[data.length];
      for (int i = 0; i < data.length; i++) {
        values[i] = (int)(data[i] * 0x1fff) & 0x3fff;
      }
      this.deconvolvedData = values;
    }
    setDeconvolved(isDeconvolved);
  }

  public AnalogDataTime(int[] data) {
    this.deconvolvedData = data;
    this.averageEnds = false;
    setDeconvolved(true);
  }

  public void checkLength(int expected) {
    if (rawData == null) {
      LengthChecker.checkLengths(deconvolvedData.length, expected);
    } else {
      LengthChecker.checkLengths(rawData.length, expected);
    }
  }

  @Override
  public Future<Void> deconvolve(DeconvolutionProxy deconvolver) {
    AnalogChannel ch = getChannel();
    Future<int[]> req = deconvolver.deconvolveAnalog(
        ch.getDacBoard(),
        ch.getDacId(),
        rawData,
        ch.getSettlingRates(),
        ch.getSettlingTimes(),
        ch.getReflectionRates(),
        ch.getReflectionAmplitudes(),
        averageEnds,
        dither
    );
    return Futures.chain(req, new Function<int[], Void>() {
      @Override
      public Void apply(int[] result) {
        deconvolvedData = result;
        setDeconvolved(true);
        return null;
      }
    });
  }

  @Override
  public int[] getDeconvolved() {
    Preconditions.checkState(isDeconvolved(), "Data has not yet been deconvolved");
    return deconvolvedData;
  }
}

File: QubitServer/src/main/scala/org/labrad/qubits/channeldata/AnalogDataTime.scala
package org.labrad.qubits.channeldata

import org.labrad.qubits.proxies.DeconvolutionProxy
import scala.concurrent.{ExecutionContext, Future}

class AnalogDataTime(data: Array[Double], isDeconvolved: Boolean, averageEnds: Boolean, dither: Boolean) extends AnalogDataBase {

  @volatile private var deconvolvedData: Array[Int] = null

  if (isDeconvolved) {
    deconvolvedData = data.map { x =>
      (x * 0x1fff).toInt & 0x3fff
    }
  }
  setDeconvolved(isDeconvolved)

  def checkLength(expected: Int): Unit = {
    LengthChecker.checkLengths(data.length, expected)
  }

  override def deconvolve(deconvolver: DeconvolutionProxy)(implicit ec: ExecutionContext): Future[Unit] = {
    val ch = getChannel()
    val req = deconvolver.deconvolveAnalog(
        ch.dacBoard,
        ch.getDacId(),
        data,
        ch.getSettlingRates(),
        ch.getSettlingTimes(),
        ch.getReflectionRates(),
        ch.getReflectionAmplitudes(),
        averageEnds,
        dither
    )
    req.map { result =>
      deconvolvedData = result
      setDeconvolved(true)
    }
  }

  override def getDeconvolved(): Array[Int] = {
    require(isDeconvolved(), "Data has not yet been deconvolved")
    deconvolvedData
  }
}

class AnalogDataTimeDacified(data: Array[Int]) extends AnalogDataBase {

  setDeconvolved(true)

  def checkLength(expected: Int): Unit = {
    LengthChecker.checkLengths(data.length, expected)
  }

  override def deconvolve(deconvolver: DeconvolutionProxy)(implicit ec: ExecutionContext): Future[Unit] = {
    sys.error("cannot deconvolve pre-dacified data")
  }

  override def getDeconvolved(): Array[Int] = {
    data
  }
}
